provider/datagen/src/baked_exporter.rs

```diff
@@ -484,12 +498,58 @@ impl DataExporter for BakedExporter {
             .map(|data| data.into_any_payload.parse::<TokenStream>().unwrap())
             .collect::<Vec<_>>();
 
+        let load_impls = match fallback_mode {
+            FallbackMode::None => {
+                lookups.iter().map(|lookup| quote! {
+                    let locale = &req.locale;
+                    if let Some(result) = #lookup {
+                        Ok(icu_provider::DataResponse {
+                            metadata: Default::default(),
+                            payload: Some(icu_provider::DataPayload::from_owned(icu_provider::prelude::zerofrom::ZeroFrom::zero_from(result))),
+                        })
+                    } else {
+                        Err(icu_provider::DataErrorKind::MissingLocale.with_req(KEY, req))
+                    }
+                }).collect::<Vec<_>>()
+            },
+            FallbackMode::Full => {
+                lookups.iter().map(|lookup| quote! {
+                    const FALLBACKER: icu_provider_adapters::fallback::LocaleFallbackerWithConfig =
+                        icu_provider_adapters::fallback::LocaleFallbackerWithConfig::from_parts(
+                            {const X: &icu_provider_adapters::fallback::provider::LocaleFallbackLikelySubtagsV1 = &singleton_fallback_likelysubtags_v1!(); X},
```

Comment:
> More opportunity for the compiler to generate duplicate data (each DataProvider impl carries its own fallback data and we rely a ton of the compiler to figure that out)

Manish said LLVM is good at this. I haven't found a place to put this data as the provider is a unit struct.

> The locale fallback is intended to be an adapter that you add to your call sites, so that there is one central function that dispatches fallback requests. If you disagree with this design, let's get alignment in a standalone issue first, not in a PR waiting to be merged.

I disagree with this. While it was initially written to work this way, fallback should be part of the data itself and the client should not use different types/call special constructors depending on whether the data needs fallback or not. I thought we had agreed on that? 

Also, doing it like this allows fallback to be more optimised, we can work directly on the key array instead of constructing `DataRequest`s and `DataResponse`s and dispatching closures on them (which is what `LocaleFallbackProvider` does).

Reply:
> (Also, hmm, icu_provider_adapters is not intended to be a dependency of component crates, so we'll need to move it into icu_provider.)

It wouldn't be a dependency of component crates, but of baked data. The `globaldata` feature would pull it in, but I think that's fine.

Reply:
Filed https://github.com/unicode-org/icu4x/issues/3504 for the fallback dependency

Reply:
Also resolved